src/braket/pulse/pulse_sequence.py

```diff
@@ -331,6 +332,56 @@ def _format_parameter_ast(
             return _FreeParameterExpressionIdentifier(parameter)
         return parameter
 
+    def _parse_json(
+        self, instr: Dict, waveforms: Dict[Waveform], frames: Dict[Frame]
+    ) -> PulseSequence:
+        """
+        Parsing a JSON input based on https://github.com/aws/amazon-braket-schemas-python/blob/main/src/braket/device_schema/pulse/native_gate_calibrations_v1.py#L26.
+
+        Args:
+            instr (Dict): The pulse instruction to parse
+            waveforms (Dict[Waveform]): The waveforms supplied for the pulse sequences.
+            frames (Dict[Frame]): A dictionary of frame objects to use.
+
+        Returns:
+            PulseSequence: The parse sequence obtain from parsing a pulse instruction.
+        """  # noqa: E501
+        if hasattr(self, f"{instr['name']}"):
+            instr_function = getattr(self, instr["name"])
+            instr_args_keys = signature(instr_function).parameters.keys()
+            instr_args = {}
+            if instr["arguments"] is not None:
+                for argument in instr["arguments"]:
+                    if argument["name"] == "waveform" and argument["name"] in instr_args_keys:
+                        instr_args[argument["name"]] = waveforms[argument["value"]]
+                    elif argument["name"] in {"qubit", "frame"}:
+                        if instr["name"] in {"barrier", "delay"}:
+                            argument_value = (
+                                [frames[argument["value"]]]
+                                if argument["name"] == "frame"
+                                else instr_args.get("qubits_or_frames", QubitSet())
+                            )
+                            (
+                                argument_value + instr_args.get("qubits_or_frames", [])
+                                if argument["name"] == "frame"
+                                else argument_value.update(QubitSet(int(argument["value"])))
+                            )
+                            instr_args["qubits_or_frames"] = argument_value
+                        else:
+                            instr_args[argument["name"]] = frames[argument["value"]]
+                    elif argument["name"] in instr_args_keys:
+                        instr_args[argument["name"]] = (
+                            float(argument["value"])
+                            if argument["type"] == "float"
+                            else FreeParameterExpression(argument["value"])
+                        )
```

Comment:
Sorry, I'm confused why we are doing these conditional checks? Why aren't we able to pass in the values directly?

Reply:
Are you asking about 
```
                            float(argument["value"])
                            if argument["type"] == "float"
                            else FreeParameterExpression(argument["value"])
```
In that case, the type is defined as  a string. The value `argument["value"]` is also a string.

Reply:
The entire block actually. Why do we need to do things like:
```
if argument["name"] == "waveform" and argument["name"] in instr_args_keys:
                        instr_args[argument["name"]] = waveforms[argument["value"]]
```
Isn't it guaranteed the service will pass in the waveform name only when it makes sense?

For the case you just described, is it possible for us to implement something generic using the argument["type"] field instead of hardcoding specifics?

Reply:
Moved to a generic argument parser function. The lookup is done for getting the waveform object from the mapping.